File: backend/app/application/services/chart_service.py
from typing import List, Dict, Any
from infrastructure.database.factory import DatabaseFactory

class ChartService:
    def __init__(self):
        self.chart_repository = DatabaseFactory.get_chart_repository()
    
    async def get_bar_chart_data(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        if filters is None:
            filters = {}
        return await self.chart_repository.get_chart_data("bar", filters)
    
    async def get_pie_chart_data(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        if filters is None:
            filters = {}
        return await self.chart_repository.get_chart_data("pie", filters)
    
    async def get_line_chart_data(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        if filters is None:
            filters = {}
        return await self.chart_repository.get_chart_data("line", filters)
    
    async def get_area_chart_data(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        if filters is None:
            filters = {}
        return await self.chart_repository.get_chart_data("area", filters)
    
    async def get_scatter_chart_data(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        if filters is None:
            filters = {}
        return await self.chart_repository.get_chart_data("scatter", filters)
    
    async def get_kpi_data(self, filters: Dict[str, Any] = None) -> Dict[str, Any]:
        if filters is None:
            filters = {}
        return await self.chart_repository.get_kpi_data(filters)
    
    async def get_chart_data_by_type(self, chart_type: str, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        if filters is None:
            filters = {}
        return await self.chart_repository.get_chart_data(chart_type, filters)

File: frontend/src/services/chartService.js
const API_BASE_URL = 'http://72.60.140.128:6002/api/v1';

class ChartService {
  getHeaders() {
    const token = localStorage.getItem('token');
    return {
      'Content-Type': 'application/json',
      ...(token && { 'Authorization': `Bearer ${token}` })
    };
  }

  async request(url, options = {}) {
    const response = await fetch(`${API_BASE_URL}${url}`, {
      headers: this.getHeaders(),
      ...options
    });
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return await response.json();
  }

  async getBarChartData(filters = {}) {
    const params = new URLSearchParams(filters);
    return this.request(`/chart/bar?${params}`);
  }

  async getPieChartData(filters = {}) {
    const params = new URLSearchParams(filters);
    return this.request(`/chart/pie?${params}`);
  }

  async getChartData(chartType, filters = {}) {
    const params = new URLSearchParams(filters);
    return this.request(`/chart/${chartType}?${params}`);
  }

  async getLineChartData(filters = {}) {
    const params = new URLSearchParams(filters);
    return this.request(`/chart/line?${params}`);
  }

  async getAreaChartData(filters = {}) {
    const params = new URLSearchParams(filters);
    return this.request(`/chart/area?${params}`);
  }

  async getScatterChartData(filters = {}) {
    const params = new URLSearchParams(filters);
    return this.request(`/chart/scatter?${params}`);
  }

  async getKPIData(filters = {}) {
    const params = new URLSearchParams(filters);
    return this.request(`/chart/kpi?${params}`);
  }
}

export const chartService = new ChartService();